File: WorldFeed/Server/src/WorldFeed.Common/Iterators/TwitterIteratorProxy.cs
namespace WorldFeed.Common.Iterators
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;

    using WorldFeed.Common.Iterators.Models;
    using WorldFeed.Common.Public.Iterators;

    public class TwitterIteratorProxy<TInput, TOutput> : TwitterIteratorProxy<TInput, TOutput, string>, ITwitterIterator<TOutput>
    {
        public TwitterIteratorProxy(ITwitterPageIterator<TInput, string> source, Func<TInput, TOutput[]> transform) : base(source, transform)
        {
        }
    }

    public class TwitterIteratorProxy<TInput, TOutput, TCursor> : ITwitterIterator<TOutput, TCursor>
    {
        private readonly ITwitterPageIterator<TInput, TCursor> source;
        private readonly Func<TInput, TOutput[]> transform;

        public TwitterIteratorProxy(ITwitterPageIterator<TInput, TCursor> source, Func<TInput, TOutput[]> transform)
        {
            this.source = source;
            this.transform = transform;
        }

        public TwitterIteratorProxy(ITwitterPageIterator<TInput, TCursor> source, Func<TInput, IEnumerable<TOutput>> transform)
        {
            this.source = source;
            this.transform = input => transform(input).ToArray();
        }

        public TCursor NextCursor { get; private set; }

        public bool Completed { get; private set; }

        public async Task<ITwitterIteratorPage<TOutput, TCursor>> NextPageAsync()
        {
            var page = await this.source.NextPageAsync().ConfigureAwait(false);
            var items = this.transform(page.Content);

            NextCursor = page.NextCursor;
            Completed = page.IsLastPage;

            return new TwitterIteratorPage<TOutput[], TOutput, TCursor>(items, NextCursor, Completed);
        }
    }
}

File: Sharebook/Client/src/app/core/Core/Iterators/TwitterIteratorProxy.ts
import {ITwitterIterator} from "../../Public/Iterators/ITwitterIterator";
import {ITwitterIteratorPage} from "../../Public/Iterators/ITwitterIteratorPage";
import {TwitterIteratorPage} from "./Models/TwitterIteratorPage";
import {ITwitterPageIterator} from "./TwitterPageIterator";

// export class TwitterIteratorProxy<TInput, TOutput> extends TwitterIteratorProxy<TInput, TOutput, string> implements ITwitterIterator<TOutput> {
//   constructor(source: ITwitterPageIterator<TInput, string>, transform: Func<TInput, TOutput[]>) {
//     super(source, transform);
//   }
// }

export class TwitterIteratorProxy<TInput, TOutput, TCursor = any> implements ITwitterIterator<TOutput, TCursor> {
  private readonly _source: ITwitterPageIterator<TInput, TCursor>;
  private readonly _transform: (tInput: TInput) => TOutput[];

  constructor(source: ITwitterPageIterator<TInput, TCursor>, transform: (tInput: TInput) => TOutput[]) {
    this._source = source;
    if (transform instanceof Array) {
      this._transform = input => transform(input);
    }
    this._transform = transform;
  }

  public nextCursor: TCursor;
  public completed: boolean;

  public async nextPageAsync(): Promise<ITwitterIteratorPage<TOutput, TCursor>> {
    let page = await this._source.nextPageAsync(); // .ConfigureAwait(false);
    let items = this._transform(page.result.Content);

    this.nextCursor = page.result.NextCursor;
    this.completed = page.result.IsLastPage;

    return new TwitterIteratorPage<TOutput[], TOutput, TCursor>(items, NextCursor, Completed);
  }
}
